test/e2e/apps/deployment.go

```diff
@@ -1705,3 +1709,146 @@ func testDeploymentSubresources(ctx context.Context, f *framework.Framework) {
+					Labels: map[string]string{
+						"test-name":                            "pod-template-hash-compatible",
+						appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+					},
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &zero,
+					Containers:                    []v1.Container{container},
+				},
+			},
+		},
+	}
+
+	_, err := c.AppsV1().ReplicaSets(ns).Create(ctx, rs, metav1.CreateOptions{})
+	framework.ExpectNoError(err, "Creating replica set %q in namespace %q", rs.Name, ns)
+
+	err = e2ereplicaset.WaitForReadyReplicaSet(ctx, c, ns, rs.Name)
+	framework.ExpectNoError(err)
+
+	// create the deployment to control the replicaset
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-deployment",
+			Namespace: ns,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &size,
+			Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"test-name": "pod-template-hash-compatible"}},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"test-name": "pod-template-hash-compatible"},
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &zero,
+					Containers:                    []v1.Container{container},
+				},
+			},
+		},
+	}
+	deployment, err := c.AppsV1().Deployments(ns).Create(ctx, d, metav1.CreateOptions{})
+	framework.ExpectNoError(err, "Creating deployment %q in namespace %q", d.Name, ns)
+	err = e2edeployment.WaitForDeploymentComplete(c, deployment)
+	framework.ExpectNoError(err)
+
+	if err := checkPodTemplateHash(ctx, 1, true, false); err != nil {
+		framework.Failf("failed checking labels after create the deployment, %v", err)
+	}
+
+	// scale should be ok with old label
+	newSize := int32(2)
+	d.Spec.Replicas = &newSize
+	deployment, err = c.AppsV1().Deployments(ns).Update(ctx, d, metav1.UpdateOptions{})
+	framework.ExpectNoError(err, "scale up deployment %q in namespace %q", d.Name, ns)
+	err = e2edeployment.WaitForDeploymentComplete(c, deployment)
+	framework.ExpectNoError(err)
+	if err := checkPodTemplateHash(ctx, 2, true, false); err != nil {
+		framework.Failf("failed checking labels after scale up the deployment, %v", err)
+	}
+
+	// update the deployment to create new labels
+	d.Spec.Replicas = &size
+	d.Spec.Template.Spec.Containers[0].Image = imageutils.GetE2EImage(imageutils.Httpd)
+	deployment, updateErr := c.AppsV1().Deployments(ns).Update(context.Background(), d, metav1.UpdateOptions{})
+	framework.ExpectNoError(updateErr, "Updating deployment %q in namespace %q", d.Name, ns)
+	err = e2edeployment.WaitForDeploymentComplete(c, deployment)
+	framework.ExpectNoError(err)
+	if err := checkPodTemplateHash(ctx, 1, true, true); err != nil {
+		framework.Failf("failed checking labels after update the deployment, %v", err)
+	}
+
+	// roll back to use the old label
+	if d.Annotations == nil {
+		d.Annotations = make(map[string]string)
+	}
+	d.Annotations[appsv1.DeprecatedRollbackTo] = "1"
```

Comment:
I suppose it is worth to test the deprecated rollback. 

Although, it is troubling that we do not have `kubectl rollout undo`  tests, as we are changing these as well and it is the primary mechanism. Would you mind adding a simple `rollout undo` test into kubectl e2e?

Reply:
:+1: to both kinds of test

Reply:
@atiratree @sftim  I'd love to help add a `rollout undo` test. However, I may need some help:

1. I used `kubectl rollout` when testing in my local cluster, but i didn't find the same operation in client-go. So I end up using this (update with annotations). Could you show me an example of doing `rollout undo` with client-go?
2. Should I add this `rolllout undo` case in this test case(pod-template-hash), or should I create another test?

Reply:
1. There is no such operation in `client-go` only in `kubectl` 
2. IMO I would keep the deployment test here only with the annotation. And then create a new test in `test/e2e/kubectl/kubectl.go` for the `rolllout undo` using `RunKubectl`.

Reply:
@atiratree I tend to create another pr to add the `rollout` test in `test/e2e/kubectl/kubectl.go` if it's ok.  These two things are not related in my opinion (update pod-template-hash and add a e2e test for rollback)

Reply:
yes separate PR is better, but it would be good to merge the kubectl tests first to make sure we are not breaking anything.

Reply:
I've created #119915 to create a simple e2e test. PTAL both prs.